theano/tensor/nnet/conv2d.py

```diff
@@ -0,0 +1,657 @@
+"""
+FIXME
+"""
+
+__docformat__ = "restructuredtext en"
+
+import logging
+
+import numpy
+
+import theano
+from theano.tensor import (as_tensor_variable, blas, get_scalar_constant_value,
+                           patternbroadcast, NotScalarConstantError)
+from theano.gof import Apply, Op
+from theano.gof import local_optimizer
+
+from theano.sandbox.cuda.basic_ops import (
+    as_cuda_ndarray_variable,
+    gpu_contiguous, gpu_from_host, host_from_gpu,
+    GpuFromHost, HostFromGpu
+    )
+from theano.sandbox.cuda import gpu_optimizer, register_opt
+from theano.sandbox.cuda.type import CudaNdarrayType
+from theano.sandbox.cuda.dnn import dnn_available, dnn_conv
+from theano.sandbox.cuda.blas import GpuCorrMM, GpuCorrMM_gradWeights, GpuCorrMM_gradInputs
+from theano.sandbox.cuda.opt import values_eq_approx_high_tol
+
+
+from theano.tensor.nnet import conv2d as cpu_conv2d
+
+imported_scipy_signal = False
+try:
+    # TODO: move these back out to global scope when they no longer
+    # cause an atexit error
+    from scipy.signal.signaltools import _valfrommode, _bvalfromboundary
+    from scipy.signal.sigtools import _convolve2d
+    imported_scipy_signal = True
+except ImportError:
+    pass
+
+_logger = logging.getLogger("theano.tensor.nnet.conv")
+
+
+def conv2d(img,
+           filters,
+           input_shape=None,
+           filter_shape=None,
+           batch_size=None,
+           border_mode='valid',
+           subsample=(1, 1),
+           filter_flip=False):
+    """
+    This function will build the symbolic graph for convolving a mini-batch of a
+    stack of 2D inputs with a set of 2D filters. The implementation is modelled
+    after Convolutional Neural Networks (CNN).
+
+    :type input: symbolic 4D tensor
+    :param input: mini-batch of feature map stacks, of shape
+        (batch size, input channels, input rows, input columns).
+        See the optional parameter ``input_shape``.
+
+    :type filters: symbolic 4D tensor
+    :param filters: set of filters used in CNN layer of shape
+        (output channels, input channels, filter rows, filter columns).
+        See the optional parameter ``filter_shape``.
+
+    :type input_shape: None, tuple/list of len 4 of int or Constant variable
+    :param input_shape: The shape of the input parameter.
+        Optional, possibly used to choose an optimal implementation.
+        You can give ``None`` for any element of the list to specify that this
+        element is not known at compile time.
```

Comment:
Something's off here. If `input_shape` is 4D, then you can't have a separate `batch_size` param. Can you delete `batch_size`?